### DISCUSSION TITLE
How to insert data in sizes of 100-200GB to a collection faster?

### DISCUSSION THREAD

**Discussion by juanandreas** [milvus-io/milvus#33887] [Q&A and General discussion]
I am currently using pymilvus 2.4.3 and my data contains sparse vector.

I am currently using client.insert() but it has a 64mb rpc limit. I split my ~115GB data table into 1750 files using pyspark to a location on databricks and upload file by file. However, this takes about 1 minute per file, which means 1750 minutes will take a whopping 29 hours!

How do I insert my data to a collection faster? I know there is spark-milvus connector but it [currently does not support sparse vector](https://github.com/zilliztech/spark-milvus/issues/18). 

I also saw there was do_bulk_insert but i kept getting an error that says 
```
    - taskID          : 450235310995975498,
    - state           : Failed,
    - row_count       : 0,
    - infos           : {'failed_reason': 'typeutil.GetDim should not invoke on sparse vector type', 'progress_percent': '0'},
    - id_ranges       : [],
    - create_ts       : 2024-06-04 15:44:37
>
```

I think this may be a bug, not sure why the bulk insert process is looking for a dimension value for sparse vector type.

**Comment by xiaofan-luan** [comment]
bulkinsertion is probably what you need. 
the next version released next week will support bulk insert sparse embeddings. https://milvus.io/docs/import-data.md
and we will also support sparse embedding soon

**Reply by juanandreas** [reply]
Could we prioritize spark milvus connector support on sparse vector as well? I would assume a majority of users that are working with this much data (and those who are working data volumes several times larger) would be utilizing spark for data manipulation before shipping it to milvus.

**Reply by juanandreas** [reply]
@xiaofan-luan Where can i find documentation on MinIO or buckets in milvus? I see the term thrown around here and there in the docs but I have yet to find conceptual documentation on how to set up a bucket in milvus. I have milvus hosted on an ec2 instance on AWS.

**Reply by yhmo** [reply]
Milvus automatically creates a bucket with the name predefined in the milvus.yaml:
```
minio:
  address: localhost # Address of MinIO/S3
  port: 9000 # Port of MinIO/S3
  ......
  bucketName: a-bucket # Bucket name in MinIO/S3
  rootPath: files # The root path where the message is stored in MinIO/S3
```

**Reply by juanandreas** [reply]
@xiaofan-luan if we are seeing:
```
"Storage backend has reached its minimum free drive threshold."
```

When trying to do bulk inserts, does this mean we have to clear object in minio manually before doing another bulk insert or do we have to resort to using s3? Is there a way to increase min io free drive threshold? Attempting to use s3 endpoints is creating issues where our milvus standalone will not start up.

**Reply by xiaofan-luan** [reply]
> @xiaofan-luan if we are seeing:
> 
> ```
> "Storage backend has reached its minimum free drive threshold."
> ```
> 
> When trying to do bulk inserts, does this mean we have to clear object in minio manually before doing another bulk insert or do we have to resort to using s3? Is there a way to increase min io free drive threshold? Attempting to use s3 endpoints is creating issues where our milvus standalone will not start up.

you are right~ milvus won't garbage collect user data and you will need to handle it manually after builk insertion done.

**Reply by juanandreas** [reply]
> > @xiaofan-luan if we are seeing:
> > ```
> > "Storage backend has reached its minimum free drive threshold."
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When trying to do bulk inserts, does this mean we have to clear object in minio manually before doing another bulk insert or do we have to resort to using s3? Is there a way to increase min io free drive threshold? Attempting to use s3 endpoints is creating issues where our milvus standalone will not start up.
> 
> you are right~ milvus won't garbage collect user data and you will need to handle it manually after builk insertion done.

@xiaofan-luan is there an api we can use to garbage collect after each bulk insertion or is s3 the best option because it's more scalable for larger datasets? Also, how long does bulk insert usually take? We have the parquet files prepared in s3, it took about 4 hours for the files to get to s3 for 125 million rows, then we ran `utility.do_bulk_insert` on the batch files. Running do_bulk_insert seems to be taking a long time.

Are there speed benchmark tests that we can look at between bulk insert vs regular inserts?

**Reply by xiaofan-luan** [reply]
1. we can add a flag to delete original file if imported finish. If imported task fail, it still have to be taken care by user since user might to retry the import task

**Reply by xiaofan-luan** [reply]
2. bulk insert speed based on how many index nodes and  datanodes you have. With enough resources it can take you in 1-2 hours.
Usually batch insertions is at least 5-10 times faster than stream insert. If you use zilliz cloud you will see your task can be accelerated more than 10 times since we have a large pool to do bulkinsertion and index build

**Reply by juanandreas** [reply]
@xiaofan-luan how do we configure how many index nodes and data nodes?

**Reply by xiaofan-luan** [reply]
if you use helm or k8s operatator, you can config node replica numbers and it's cpu request/limit

**Reply by juanandreas** [reply]
@xiaofan-luan is bulk insert equivalent to stream insert if our milvus database is hosted on a single machine (ec2 instance)?  Is bulk insert intended for a horizontally scaled milvus db? If we use ziliz managed service, will we be able to configure index/data nodes for  efficient bulk inserts?

**Reply by bigsheeper** [reply]
> @xiaofan-luan is bulk insert equivalent to stream insert if our milvus database is hosted on a single machine (ec2 instance)? Is bulk insert intended for a horizontally scaled milvus db? If we use ziliz managed service, will we be able to configure index/data nodes for efficient bulk inserts?

@juanandreas 
1. Bulk insert parses files and writes to the binlog directly, making it faster than streaming inserts.
2. Yep, the more datanodes and indexnodes you have, the quicker the bulk insert will go.
3. You don't need to, for Zilliz Cloud, we've implemented many optimizations for bulk insert, resulting in a much faster process.
